src/test/java/htsjdk/samtools/CachingBAMFileIndexTest.java

```diff
@@ -0,0 +1,93 @@
+     * this tests for a bug that was hidden on human data pre-hg38 due to a weird interaction between WeakHashMap Integer interning
+     */
+    @Test
+    public void testCachingBamFileIndextest() throws IOException {
+        try(final CachingBAMFileIndex index = getIndexWith200Contigs()) {
+            Assert.assertNotNull(index.getQueryResults(1));
+            System.gc();
+            //contig 1 is never garbage collected because Integer(1) is interned by the jvm and never garbage collectable
+            Assert.assertNotNull(index.getQueryResults(1));
+
+            Assert.assertNotNull(index.getQueryResults(128));
+            System.gc();
+            //contig 128 is garbage collected and hits the bug because 128+ is not interned
+            Assert.assertNotNull(index.getQueryResults(128));
+        }
+    }
+
+    private static CachingBAMFileIndex getIndexWith200Contigs() throws IOException {
+        List<SAMSequenceRecord> contigs = IntStream.range(1, 200)
+                .mapToObj(i -> new SAMSequenceRecord(String.valueOf(i), 1000))
+                .collect(Collectors.toList());
+        final SAMSequenceDictionary dict = new SAMSequenceDictionary(contigs);
+        final SAMFileHeader header = new SAMFileHeader(dict);
+        header.setSortOrder(SAMFileHeader.SortOrder.coordinate);
+
+        final SAMFileWriterFactory writerFactory = new SAMFileWriterFactory().setCreateIndex(true);
+        final File outBam = File.createTempFile("tmp", ".bam");
+        try(final SAMFileWriter writer = writerFactory.makeWriter(header, true, outBam, null)){
+            IntStream.range(1,200).mapToObj(i -> {
+                final SAMRecord record = new SAMRecord(header);
+                record.setReadName("name" + i);
+                record.setReferenceName(String.valueOf(i));
+                record.setReadUnmappedFlag(false);
+                record.setAlignmentStart(1);
+                record.setCigarString("20M");
+                return record;
+            }).forEach(writer::addAlignment);
+        }
+
+        final File indexFile = new File(outBam.getParent(), IOUtil.basename(outBam) + BAMIndex.BAMIndexSuffix);
+        indexFile.deleteOnExit();
+        outBam.deleteOnExit();
+        return new CachingBAMFileIndex(indexFile, dict);
+    }
+
+    @Test
+    public void testCacheHitsAndMissesTheExpectedNumberOfTimes() throws IOException {
+        try(final CachingBAMFileIndex index = getIndexWith200Contigs()) {
+            index.getQueryResults(1);
+            assertCacheStats(index, 0, 1);
+
+            index.getQueryResults(1);
+            assertCacheStats(index, 1, 1);
+
+            index.getQueryResults(150);
+            assertCacheStats(index, 1, 2);
+
+            index.getQueryResults(150);
+            assertCacheStats(index,2,2);
+
+            index.getQueryResults(150);
+            assertCacheStats(index,3,2);
+
+            index.getQueryResults(1);
+            assertCacheStats(index, 3, 3);
```

Comment:
Add one additional query on chromosome 1 to make sure it gets put back in the cache.